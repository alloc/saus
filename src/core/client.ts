import fs from 'fs'
import MagicString from 'magic-string'
import {
  flattenCallChain,
  getBabelProgram,
  isChainedCall,
  NodePath,
  removeSSR,
  resolveReferences,
  t,
} from '../babel'
import { renderIdentRE } from '../plugins/render'
import type { RouteParams } from './routes'

/** A generated client module */
export interface Client {
  id: string
  code: string
  map?: ExistingRawSourceMap | null
}

interface ExistingRawSourceMap {
  file?: string
  mappings: string
  names: string[]
  sourceRoot?: string
  sources: string[]
  sourcesContent?: string[]
  version: number
}

/** JSON state provided by the renderer and made available to the client */
export type ClientState = Record<string, any> & {
  rootId?: string
  routePath: string
  routeParams: RouteParams
  error?: any
}
/**
 * Client fragments represent a portion of a render module,
 * which are pieced together by the SSR bundle based on
 * which page URL is being rendered.
 */
export interface ClientFunction {
  /** Route path for page matching */
  route?: string
  /** Character offset where the function is found within the source file */
  start: number
  /** The function implementation */
  function: string
  /** Referenced variables outside the function */
  referenced: string[]
}

export interface RenderFunction extends ClientFunction {
  didRender?: ClientFunction
}

export type ClientFunctions = {
  filename: string
  beforeRender: ClientFunction[]
  render: RenderFunction[]
}

export function mapClientFunctions<T>(
  functions: ClientFunctions,
  map: (fn: ClientFunction) => T
): T[] {
  const mapped: T[] = functions.beforeRender.map(map)
  functions.render.forEach(renderFn => {
    mapped.push(map(renderFn))
    if (renderFn.didRender) {
      mapped.push(map(renderFn.didRender))
    }
  })
  return mapped
}

export function extractClientFunctions(filename: string): ClientFunctions {
  const rawSource = fs.readFileSync(filename, 'utf8')
  const source = new MagicString(rawSource, {
    filename,
    indentExclusionRanges: [],
  })

  const beforeRenderFns: NodePath<t.ArrowFunctionExpression>[] = []
  const renderFns: NodePath<t.ArrowFunctionExpression>[] = []
  const didRenderFns: Record<number, NodePath<t.ArrowFunctionExpression>> = {}

  const program = getBabelProgram(rawSource, filename)
  program.traverse({
    CallExpression(callPath) {
      const callee = callPath.get('callee')
      if (callee.isIdentifier() && renderIdentRE.test(callee.node.name)) {
        const start = callPath.node.start!
        callPath.traverse({
          Identifier(path) {
            const { node, parentPath } = path

            // Parse the `render(...)` call
            if (node.start === start && parentPath.isCallExpression()) {
              parentPath.get('arguments').find(arg => {
                if (arg.isArrowFunctionExpression()) {
                  renderFns.push(arg)
                  return true
                }
              })
              path.stop()
            }

            // Parse the `.then(...)` call
            else if (isChainedCall(parentPath)) {
              const callChain = flattenCallChain(parentPath)
              if (callChain[0].node.start === start) {
                const thenCall =
                  parentPath.parentPath as NodePath<t.CallExpression>

                const thenProperty = thenCall
                  .get('callee')
                  .get('property') as NodePath

                if (!thenProperty.isIdentifier({ name: 'then' })) {
                  return
                }

                thenCall.get('arguments').find(arg => {
                  if (arg.isArrowFunctionExpression()) {
                    didRenderFns[start] = arg
                    return true
                  }
                })
              }
            }
          },
        })
      } else if (callee.isIdentifier({ name: 'beforeRender' })) {
        for (const arg of callPath.get('arguments')) {
          if (arg.isArrowFunctionExpression()) {
            beforeRenderFns.push(arg)
            break
          }
        }
      }
    },
  })

  const getSource = (path: NodePath) => {
    const { start, end } = path.node
    if (start == null || end == null) {
      throw Error('Missing node start/end')
    }
    return source.slice(start, end)
  }

  const defineClientFunction = (
    fn: NodePath<t.ArrowFunctionExpression>
  ): ClientFunction => {
    const parentArgs = fn.parentPath.get('arguments') as NodePath[]
    const routeArg =
      parentArgs[0] !== fn
        ? (parentArgs[0] as NodePath<t.StringLiteral>)
        : undefined

    const referenced = resolveReferences(
      fn.get('body'),
      path => !path.isDescendant(fn.parentPath)
    ).map(getSource)

    return {
      route: routeArg?.node.value,
      start: fn.parentPath.node.start!,
      function: getSource(fn),
      referenced,
    }
  }

  return {
    filename,
    beforeRender: beforeRenderFns.map(defineClientFunction),
    render: renderFns.map(path => {
      const fn = defineClientFunction(path) as RenderFunction
      const didRenderFn = didRenderFns[fn.start]
      if (didRenderFn) {
        didRenderFn.get('body').traverse(removeSSR(source))
        fn.didRender = defineClientFunction(didRenderFn)
      }
      return fn
    }),
  }
}

export type ClientImports = {
  [source: string]: string | (string | [name: string, alias: string])[]
}

export function defineClient(description: ClientDescription) {
  return description
}

export interface ClientDescription {
  /**
   * Define `import` statements to be included.
   *
   * The keys are modules to import from, and the values are either the
   * identifier used for the default export or an array of identifiers
   * used for named exports.
   */
  imports: ClientImports
  /**
   * Hydration code to run on the client.
   *
   * Executed inside a function with this type signature:
   *
   *     async (content: unknown, request: RenderRequest) => void
   *
   * Custom imports are available as well.
   */
  onHydrate: string
}
