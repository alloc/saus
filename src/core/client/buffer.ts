// Simple wrapper around TextDecoder to help provide HTTP helpers
// that work in all environments.

export class Buffer {
  protected constructor(readonly buffer: ArrayBuffer) {}

  get length() {
    return this.buffer.byteLength
  }

  static isBuffer(buf: any): buf is Buffer {
    return buf instanceof Buffer
  }

  static from(data: ArrayBuffer) {
    return new Buffer(data)
  }

  toString(encoding?: string): string {
    const decoder = new TextDecoder(encoding)
    return decoder.decode(this.buffer)
  }
}

export function unwrapBuffer(buf: Buffer) {
  return buf // Client buffers remain unchanged.
}
