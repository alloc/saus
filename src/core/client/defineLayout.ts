import type { RouteLayout } from '@/runtime/layouts'

export function defineLayout(layout: RouteLayout) {
  return layout
}
