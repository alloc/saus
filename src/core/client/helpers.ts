export { describeHead } from './head'
export { hydrate } from './hydrate'
export { preloadModules } from './preloadModules'
export { importStateModules } from './stateModules/loader'
