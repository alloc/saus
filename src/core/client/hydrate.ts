import type { CommonClientProps, RenderRequest } from '../core'
import { globalCache } from '../runtime/cache'
import type { RouteLayout } from '../runtime/layouts'

export type Hydrator<RenderResult = any> = (
  root: HTMLElement,
  content: RenderResult,
  request: RenderRequest
) => Promise<void> | void

export interface RouteClient {
  hydrate: Hydrator
  layout: Pick<RouteLayout, 'render' | 'clientHooks'>
  routeModule: object
}

export const defineHydrator = <RenderResult = any>(
  hydrate: Hydrator<RenderResult>
) => hydrate

export async function hydrate(
  { hydrate, layout, routeModule }: RouteClient,
  props: CommonClientProps,
  root: HTMLElement
) {
  // Cache page props in global cache.
  const path = location.pathname
  globalCache.loaded[path] = [props] // TODO: should this expire?

  const req: RenderRequest = {
    path,
    query: location.search.slice(1),
    params: props.routeParams,
    module: routeModule,
    props,
  }

  const content = await layout.render(req)

  const { beforeHydrate, afterHydrate } = layout.clientHooks || {}
  if (beforeHydrate) {
    await beforeHydrate(req, root)
  }

  await hydrate(root, content, req)
  saus.hydrated = true

  if (afterHydrate) {
    await afterHydrate(req, root)
  }
}
