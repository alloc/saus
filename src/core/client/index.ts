import './context'

export * from './api'
export * from './baseUrl'
export * from './buffer'
export * from './defineLayout'
export * from './head'
export * from './http/get'
export * from './hydrate'
export * from './index.dev'
export * from './loadPageState'
export * from './pageClient'
export * from './prependBase'
export { default as routes } from './routes'
export * from './types'
