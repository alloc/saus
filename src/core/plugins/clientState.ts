import { babel, getBabelConfig, NodePath, resolveReferences, t } from '../babel'
import { SourceMap } from '../node/sourceMap'
import { Plugin } from '../vite'

const includeRE = /\.m?[tj]sx?$/

/**
 * Transform `defineStateModule` calls for client-side use.
 */
export function clientStatePlugin(): Plugin {
  return {
    name: 'saus:state:client',
    enforce: 'pre',
    async transform(code, id, opts) {
      if (opts?.ssr) {
        return
      }
      if (!includeRE.test(id)) {
        return // Unsupported file type
      }
      if (id.includes('/saus/src/')) {
        return // Saus core modules
      }
      if (/\bdefineStateModule\b/.test(code)) {
        const parsed = await babel.parseAsync(code, getBabelConfig(id))
        if (!parsed) {
          return
        }

        const preserved = new Set<NodePath<t.Statement>>()

        babel.traverse(parsed, {
          ExportDeclaration(path) {
            preserved.add(path)
          },
          CallExpression(path) {
            const callee = path.get('callee')
            if (callee.isIdentifier({ name: 'defineStateModule' })) {
              const args = path.get('arguments')
              for (let i = 1; i < args.length; i++) {
                // Remove all arguments except the first.
                args[i].remove()
              }
            }
            // FIXME: Be smarter about which onLoad calls we preserve.
            else if (callee.toString().endsWith('.onLoad')) {
              preserved.add(path.getStatementParent()!)
              path.skip()
            }
          },
        })

        resolveReferences(Array.from(preserved)).forEach(path => {
          preserved.add(path)
        })

        const transformer: babel.Visitor = {
          Program(path) {
            const stmts = new Set(
              Array.from(preserved).sort(
                (a, b) => a.node.start! - b.node.start!
              )
            )

            for (const stmt of stmts) {
              path.node.body.push(stmt.node)
            }
          },
        }

        const transformed = babel.transformSync('', {
          plugins: [{ visitor: transformer }],
          sourceMaps: true,
        }) as {
          code: string
          map: SourceMap
        }

        transformed.map.sources = [id]
        transformed.map.sourcesContent = [code]

        return transformed
      }
    },
  }
}
