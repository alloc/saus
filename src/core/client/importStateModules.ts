import { prependBase } from '@/client/prependBase'
import { globalCache } from '@/runtime/cache'
import saus from './context'

/**
 * Note: This doesn't support on-demand state modules, so it should
 * only be used for state modules coupled to the page's route, which
 * means the state module is loaded unconditionally when the page
 * is requested.
 */
export function importStateModules(...args: any[]) {
  return Promise.all(
    args.map(async cacheKey => {
      const [state] = await globalCache.access(cacheKey, async () => {
        const stateUrl = prependBase(saus.stateModuleBase + cacheKey + '.js')
        await import(/* @vite-ignore */ stateUrl)

        // Skip updating the cache and use the cache entry that was
        // injected by the module we just imported.
        return Symbol.for('skip')
      })
      return state
    })
  )
}
