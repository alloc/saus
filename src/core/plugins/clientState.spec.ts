import endent from 'endent'
import { expect, test } from 'vitest'
import { SourceMap } from '../node/sourceMap'
import { clientStatePlugin } from './clientState'

const code = endent`
  import { defineStateModule } from "saus/client"
  import someClientLib from "some-client-lib"
  import anotherClientLib from "another-client-lib"
  import { fetch } from "node-fetch"

  export const foo = defineStateModule('foo', () => fetch('/foo'))

  const clientSideEffect = (args, state, expiresAt) => {
    someClientLib(state)
  }

  foo.onLoad(clientSideEffect)

  export { anotherClientLib }
`

test('clientStatePlugin', async () => {
  const plugin = clientStatePlugin()
  const transform = plugin.transform as (
    code: string,
    id: string,
    opts: { ssr: boolean }
  ) => Promise<{
    code: string
    map: SourceMap
  }>

  const result = await transform(code, 'state.ts', { ssr: false })
  if (!result || typeof result == 'string') {
    throw new Error('Expected result to be an object')
  }

  // result.code += toInlineSourceMap(result.map)

  expect(result).toMatchSnapshot()
  expect(result.map).not.toBe(null)
})
