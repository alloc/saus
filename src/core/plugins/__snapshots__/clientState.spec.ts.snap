// Vitest Snapshot v1

exports[`clientStatePlugin 1`] = `
{
  "ast": null,
  "code": "import { defineStateModule } from \\"saus/client\\";
import someClientLib from \\"some-client-lib\\";
import anotherClientLib from \\"another-client-lib\\";
export const foo = defineStateModule('foo');

const clientSideEffect = (args, state, expiresAt) => {
  someClientLib(state);
};

foo.onLoad(clientSideEffect);
export { anotherClientLib };",
  "externalDependencies": Set {},
  "map": {
    "file": undefined,
    "mappings": "AAAA,SAASA,iBAAT,QAAkC,aAAlC;AACA,OAAOC,aAAP,MAA0B,iBAA1B;AACA,OAAOC,gBAAP,MAA6B,oBAA7B;AAGA,OAAO,MAAMC,GAAG,GAAGH,iBAAiB,CAAC,KAAD,CAA7B;;AAEP,MAAMI,gBAAgB,GAAG,CAACC,IAAD,EAAOC,KAAP,EAAcC,SAAd,KAA4B;EACnDN,aAAa,CAACK,KAAD,CAAb;AACD,CAFD;;AAIAH,GAAG,CAACK,MAAJ,CAAWJ,gBAAX;AAEA,SAASF,gBAAT",
    "names": [
      "defineStateModule",
      "someClientLib",
      "anotherClientLib",
      "foo",
      "clientSideEffect",
      "args",
      "state",
      "expiresAt",
      "onLoad",
    ],
    "sourceRoot": undefined,
    "sources": [
      "state.ts",
    ],
    "sourcesContent": [
      "import { defineStateModule } from \\"saus/client\\"
import someClientLib from \\"some-client-lib\\"
import anotherClientLib from \\"another-client-lib\\"
import { fetch } from \\"node-fetch\\"

export const foo = defineStateModule('foo', () => fetch('/foo'))

const clientSideEffect = (args, state, expiresAt) => {
  someClientLib(state)
}

foo.onLoad(clientSideEffect)

export { anotherClientLib }",
    ],
    "version": 3,
  },
  "metadata": {},
  "options": {
    "assumptions": {},
    "babelrc": false,
    "browserslistConfigFile": false,
    "cloneInputAst": true,
    "configFile": false,
    "cwd": "/Users/alec/dev/alloc/saus",
    "envName": "test",
    "generatorOpts": {
      "auxiliaryCommentAfter": undefined,
      "auxiliaryCommentBefore": undefined,
      "comments": true,
      "compact": "auto",
      "filename": undefined,
      "minified": undefined,
      "retainLines": undefined,
      "shouldPrintComment": undefined,
      "sourceFileName": "unknown",
      "sourceMaps": true,
      "sourceRoot": undefined,
    },
    "parserOpts": {
      "plugins": [],
      "sourceFileName": undefined,
      "sourceType": "module",
    },
    "passPerPreset": false,
    "plugins": [
      Plugin {
        "externalDependencies": [],
        "generatorOverride": undefined,
        "key": "base$0",
        "manipulateOptions": undefined,
        "options": {},
        "parserOverride": undefined,
        "post": undefined,
        "pre": undefined,
        "visitor": {
          "Program": {
            "enter": [
              [Function],
            ],
          },
          "_exploded": true,
          "_verified": true,
        },
      },
    ],
    "presets": [],
    "root": "/Users/alec/dev/alloc/saus",
    "rootMode": "root",
    "sourceMaps": true,
    "targets": {},
  },
  "sourceType": "module",
}
`;
